### PR TITLE
VERSION keyword for Earthfile

### PR DESCRIPTION
**Pull request by alexcb**
- currently the VERSION is optional, but it will allow us to feature-flip
options at a future point.

Part of https://github.com/earthly/earthly/issues/991

Signed-off-by: Alex Couture-Beil <alex@earthly.dev>

### CODE REVIEW

File: ast/tests/build-arg.ast.json [hunk 1]
```diff
@@ -1,4 +1,8 @@
 {
+  "Version": {
+    "major": 0,
+    "minor": 0
+  },
```
Comment (vladaionescu) [inline ast/tests/build-arg.ast.json:5]: I think it's better if this is left out when it's not set. It's less error prone when the JSON AST matches exactly to what is actually visible in text form.

File: ast/spec/earthfile.go [hunk 2]
```diff
@@ -2,6 +2,7 @@ package spec
 
 // Earthfile is the AST representation of an Earthfile.
 type Earthfile struct {
+	Version        Version
```
Comment (vladaionescu) [inline ast/spec/earthfile.go:5]: `json:"version"` (lower-case, to match the other fields)

Comment (vladaionescu) [inline ast/spec/earthfile.go:5]: Should probably also be optional (`*Version` and `omitempty`)

File: ast/version.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+				EndColumn:   lineWidth,
+			}
+		}
+
+		return &version, nil
+	}
+
+	// No version was found
+	return nil, nil
+}
```
Comment (vladaionescu) [inline ast/version.go:108]: ~This could use some tests - it's nicely isolated.~

NVM - seems like you have some nice integration tests.

File: earthfile2llb/earthfile2llb.go [hunk 4]
```diff
@@ -97,6 +99,16 @@ func Earthfile2LLB(ctx context.Context, target domain.Target, opt ConvertOpt) (m
 	if err != nil {
 		return nil, errors.Wrapf(err, "resolve build context for target %s", target.String())
 	}
+
+	ftrs, err := features.GetFeatures(bc.Earthfile.Version)
+	if err != nil {
+		return nil, errors.Wrapf(err, "resolve build context for target %s", target.String())
```
Comment (vladaionescu) [inline earthfile2llb/earthfile2llb.go:105]: (You'll need to also define a `bc.Earthfile.Version.String()` or similar)

```suggestion
		return nil, errors.Wrapf(err, "resolve feature set for version %s for target %s", bc.Earthfile.Version.String(), target.String())
```

File: earthfile2llb/earthfile2llb.go [hunk 5]
```diff
@@ -97,6 +99,16 @@ func Earthfile2LLB(ctx context.Context, target domain.Target, opt ConvertOpt) (m
...
+	}
+
+	if ftrs.ReferencedSaveOnly {
+		fmt.Printf("TODO feature-flip referenced save artifact as local feature in a future PR.\n")
```
Comment (vladaionescu) [inline earthfile2llb/earthfile2llb.go:109]: Somewhat unrelated, but this will apply to both save artifact and save image, right?

File: examples/tests/version/Earthfile [hunk 6]
```diff
@@ -0,0 +1,46 @@
+ARG DOCKERHUB_AUTH=true
+FROM ../../..+earthly-integration-test-base --DOCKERHUB_AUTH=$DOCKERHUB_AUTH
+
+IMPORT .. AS tests
```
Comment (vladaionescu) [inline examples/tests/version/Earthfile:4]: 🆒

File: examples/tests/version/Earthfile [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    BUILD +test-single-line
+    BUILD +test-single-line-with-args
+    BUILD +test-multi-line
+    BUILD +test-multi-line-with-comment
+    BUILD +test-multi-line-with-comment2
+    BUILD +test-multi-line-with-comment3
+    BUILD +test-multi-line-with-comment4
+    BUILD +test-multi-line-with-args
+    BUILD +test-multi-line-with-args2
+    BUILD +test-multi-line-with-newline
```
Comment (vladaionescu) [inline examples/tests/version/Earthfile:46]: Some more edge-cases:
* A comment before version, possibly with whitespace in-between
* Whitespace, then `VERSION`
* `VERSION`, then EOF (no new line)

File: features/features.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+	if version == nil {
+		return &ftrs, nil
+	}
+
+	if version.Args == nil {
+		return nil, errUnexpectedArgs
+	}
+
+	parsedArgs, err := flagutil.ParseArgs("VERSION", &ftrs, version.Args)
```
Comment (vladaionescu) [inline features/features.go:36]: Could this be made to ignore feature flags that we don't recognize? Maybe only show a warning in such cases?

File: features/features.go [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+		return nil, errors.Wrapf(err, "failed to parse major version %q", majorAndMinor[0])
+	}
+	minor, err := strconv.Atoi(majorAndMinor[1])
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse minor version %q", majorAndMinor[1])
+	}
+
+	// TODO depending on versions here, flip on/off features.
+	_ = major
+	_ = minor
```
Comment (vladaionescu) [inline features/features.go:60]: Might be useful to carry a string form for this, to display in error messages if needed?

**Review by vladaionescu — COMMENTED** [review]
Cool stuff

File: earthfile2llb/earthfile2llb.go (hunk 5 above)
Comment (alexcb) [reply earthfile2llb/earthfile2llb.go:109]: That was my intention, a single bool to flip both of those on and off.

File: examples/tests/version/Earthfile (hunk 7 above)
Comment (alexcb) [reply examples/tests/version/Earthfile:57]: Good ideas.

Having an empty Earthfile is problematic, as I don't think I can run it without a `+test` target to run; however, I discovered I can `IMPORT` an empty Earthfile simply containing a `VERSION 0.5`, and that's passing.

File: features/features.go (hunk 8 above)
Comment (alexcb) [reply features/features.go:36]: I'm unsure if issuing a warning is a good idea here, 

What about a case where a user wants to make use of a certain feature but that feature doesn't exist? for example:

    VERSION --enable-foo 0.5

If we ran this on an earlier version of earthly, then it would just print a warning and then likely fail later on when foo isn't enabled.

Can you think of an example where this ignore unknown flags behaviour would help a user?

File: earthfile2llb/earthfile2llb.go (hunk 4 above)
Comment (alexcb) [reply earthfile2llb/earthfile2llb.go:105]: Currently the structs in the ast/spec package only have JSON tags and no methods (e.g. see https://github.com/earthly/earthly/blob/version/ast/spec/earthfile.go ). A quick idea is to just use `%v` and `bc.Earthfile.Version.Args`, or we could just marshal the str via `json.marshal`?

File: examples/tests/version/Earthfile (hunk 7 above)
Comment (vladaionescu) [reply examples/tests/version/Earthfile:57]: You could try to run `+base`

File: features/features.go (hunk 8 above)
Comment (vladaionescu) [reply features/features.go:36]: Not a strong opinion. I was thinking of a case where multiple people on the team are using varying versions of Earthly and only some have the version with that feature flag. If the feature flag errors out on a version of Earthly that doesn't support it, then the users wouldn't be able to use any target from that Earthfile.

I'm not convinced either that this behavior is useful. So maybe it's better to return an error if we don't recognize it.

File: examples/tests/version/Earthfile (hunk 7 above)
Comment (alexcb) [reply examples/tests/version/Earthfile:57]: of course! I forget that can be run from the command line :+1:

File: features/features.go (hunk 8 above)
Comment (alexcb) [reply features/features.go:36]: Maybe that's where `--strict` could come into play? I'm in favour of leaving it as-is until a use-case comes up requiring it.

Comment (vladaionescu) [reply features/features.go:36]: Yeah, less is more. As-is is best.

File: features/features.go (hunk 9 above)
Comment (alexcb) [reply features/features.go:98]: I added in `Version()` and `String()` methods on the `Features` struct to make it easier to get a string version for debugging/errors.

**Review by vladaionescu — APPROVED** [review]
